File: web/src/services/api/visions.ts
import { http } from "./client";
import { ENDPOINTS } from "./endpoints";
import type { PersonSummary } from "./types/common";
import type { Task } from "./tasks";
import type { UUID } from "@/types/primitive";
import type { ListResponse } from "@/types/pagination";
// Types local to visions
export interface Vision {
  id: UUID;
  name: string;
  description?: string | null;
  area_id?: UUID | null;
  status: string;
  stage: number;
  experience_points: number;
  experience_rate_per_hour: number | null;
  created_at: string;
  people?: PersonSummary[];
  // Optional aggregated minutes for display (if backend adds later)
  total_actual_effort?: number | null;
}

export interface VisionCreate {
  name: string;
  description?: string;
  area_id?: UUID | null;
  person_ids?: UUID[];
  status?: string;
  experience_rate_per_hour?: number | null;
}

export interface VisionUpdate {
  name?: string;
  description?: string;
  status?: string;
  area_id?: UUID | null;
  person_ids?: UUID[];
  experience_rate_per_hour?: number | null;
}

export interface VisionWithTasks extends Vision {
  tasks: Task[];
}

export interface VisionStatsResponse {
  total_tasks: number;
  completed_tasks: number;
  in_progress_tasks: number;
  todo_tasks: number;
  completion_percentage: number;
  total_estimated_effort?: number | null;
  total_actual_effort?: number | null;
}

export interface VisionExperienceRateUpdatePayload {
  id: UUID;
  experience_rate_per_hour: number | null;
}

interface VisionListMeta {
  status_filter?: string | null;
}

export type VisionListResponse = ListResponse<Vision, VisionListMeta>;

export const visionsApi = {
  async getAll(
    statusFilter?: string,
    page: number = 1,
    size: number = 100,
  ): Promise<VisionListResponse> {
    return http.get<VisionListResponse>(ENDPOINTS.VISIONS.BASE, {
      status_filter: statusFilter,
      page,
      size,
    });
  },

  async getById(id: UUID): Promise<Vision> {
    return http.get<Vision>(ENDPOINTS.VISIONS.BY_ID(id));
  },

  async getWithTasks(id: UUID): Promise<VisionWithTasks> {
    return http.get<VisionWithTasks>(ENDPOINTS.VISIONS.WITH_TASKS(id));
  },

  async create(vision: VisionCreate): Promise<Vision> {
    return http.post<Vision>(ENDPOINTS.VISIONS.BASE, vision);
  },

  async update(id: UUID, vision: VisionUpdate): Promise<Vision> {
    return http.patch<Vision>(ENDPOINTS.VISIONS.BY_ID(id), vision);
  },

  async delete(id: UUID): Promise<void> {
    return http.delete<void>(ENDPOINTS.VISIONS.BY_ID(id));
  },

  async addExperience(id: UUID, experiencePoints: number): Promise<Vision> {
    return http.post<Vision>(ENDPOINTS.VISIONS.ADD_EXPERIENCE(id), {
      experience_points: experiencePoints,
    });
  },

  async harvest(id: UUID): Promise<Vision> {
    return http.post<Vision>(ENDPOINTS.VISIONS.HARVEST(id));
  },

  async getStats(id: UUID): Promise<VisionStatsResponse> {
    return http.get<VisionStatsResponse>(ENDPOINTS.VISIONS.STATS(id));
  },

  async recomputeEfforts(
    id: UUID,
  ): Promise<{ vision_id: UUID; recomputed_roots: UUID[] }> {
    return http.post<{ vision_id: UUID; recomputed_roots: UUID[] }>(
      ENDPOINTS.VISIONS.RECOMPUTE_EFFORTS(id),
    );
  },

  async syncExperience(id: UUID): Promise<Vision> {
    return http.post<Vision>(ENDPOINTS.VISIONS.SYNC_EXPERIENCE(id));
  },

  async bulkUpdateExperienceRates(
    items: VisionExperienceRateUpdatePayload[],
  ): Promise<Vision[]> {
    const updated = await Promise.all(
      items.map((item) =>
        visionsApi.update(item.id, {
          experience_rate_per_hour: item.experience_rate_per_hour,
        }),
      ),
    );
    return updated;
  },
};

File: src/lifeos_web/routers/visions.py
"""Vision endpoints used by the local planning UI."""

from __future__ import annotations

import math
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from lifeos_cli.db.services import visions as vision_services
from lifeos_web.deps import get_db_session
from lifeos_web.schemas import ListResponse, Pagination, VisionCreate, VisionUpdate
from lifeos_web.serialization import to_jsonable, to_jsonable_dict

router = APIRouter(prefix="/visions", tags=["visions"])
SessionDep = Annotated[AsyncSession, Depends(get_db_session)]


def _vision_payload(vision: object, *, include_tasks: bool = False) -> dict[str, object]:
    """Serialize a vision for the Web UI."""
    payload: dict[str, object] = {
        "id": str(vision.id),
        "name": vision.name,
        "description": vision.description,
        "area_id": str(vision.area_id) if vision.area_id else None,
        "status": vision.status,
        "stage": vision.stage,
        "experience_points": vision.experience_points,
        "experience_rate_per_hour": vision.experience_rate_per_hour,
        "created_at": to_jsonable(vision.created_at),
        "people": to_jsonable(vision.people),
    }
    if include_tasks:
        payload["tasks"] = to_jsonable(vision.tasks)
    return payload


@router.get("/", response_model=ListResponse)
async def list_visions(
    session: SessionDep,
    page: int = Query(1, ge=1),
    size: int = Query(100, ge=1, le=500),
    status_filter: str | None = None,
) -> ListResponse:
    """List active visions for selectors and planning cards."""
    rows = await vision_services.list_visions(
        session,
        status=status_filter,
        limit=size,
        offset=(page - 1) * size,
    )
    items = [_vision_payload(row) for row in rows]
    return ListResponse(
        items=items,
        pagination=Pagination(
            page=page,
            size=size,
            total=len(items),
            pages=math.ceil(len(items) / size) if size > 0 else 0,
        ),
        meta={"status_filter": status_filter},
    )


@router.get("/{vision_id}")
async def get_vision(vision_id: UUID, session: SessionDep) -> dict[str, object]:
    """Load one vision."""
    vision = await vision_services.get_vision(session, vision_id=vision_id)
    if vision is None:
        raise HTTPException(status_code=404, detail=f"Vision {vision_id} was not found")
    return _vision_payload(vision)


@router.post("/")
async def create_vision(
    payload: VisionCreate,
    session: SessionDep,
) -> dict[str, object]:
    """Create a vision."""
    try:
        vision = await vision_services.create_vision(
            session,
            name=payload.name,
            description=payload.description,
            status=payload.status,
            area_id=payload.area_id,
            experience_rate_per_hour=payload.experience_rate_per_hour,
            person_ids=payload.person_ids,
        )
    except (LookupError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _vision_payload(vision, include_tasks=True)


@router.put("/{vision_id}")
async def replace_vision(
    vision_id: UUID,
    payload: VisionUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Frontend-compatible alias for partial vision update."""
    return await update_vision(vision_id=vision_id, payload=payload, session=session)


@router.patch("/{vision_id}")
async def update_vision(
    vision_id: UUID,
    payload: VisionUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Update a vision."""
    fields = payload.model_fields_set
    clear_description = "description" in fields and payload.description is None
    clear_experience_rate = (
        "experience_rate_per_hour" in fields and payload.experience_rate_per_hour is None
    )
    clear_area = "area_id" in fields and payload.area_id is None
    clear_people = "person_ids" in fields and payload.person_ids == []
    try:
        vision = await vision_services.update_vision(
            session,
            vision_id=vision_id,
            name=payload.name,
            description=payload.description,
            clear_description=clear_description,
            status=payload.status,
            area_id=payload.area_id,
            clear_area=clear_area,
            experience_rate_per_hour=payload.experience_rate_per_hour,
            clear_experience_rate=clear_experience_rate,
            person_ids=payload.person_ids,
            clear_people=clear_people,
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _vision_payload(vision)


@router.delete("/{vision_id}", status_code=204)
async def delete_vision(
    vision_id: UUID,
    session: SessionDep,
) -> None:
    """Soft-delete a vision."""
    try:
        await vision_services.delete_vision(session, vision_id=vision_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/{vision_id}/with-tasks")
async def get_vision_with_tasks(vision_id: UUID, session: SessionDep) -> dict[str, object]:
    """Load a vision with active tasks."""
    try:
        vision = await vision_services.get_vision_with_tasks(session, vision_id=vision_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _vision_payload(vision)


@router.get("/{vision_id}/stats")
async def get_vision_stats(vision_id: UUID, session: SessionDep) -> dict[str, object]:
    """Load vision task stats."""
    try:
        stats = await vision_services.get_vision_stats(session, vision_id=vision_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return to_jsonable_dict(stats)


@router.post("/{vision_id}/add-experience")
async def add_experience(
    vision_id: UUID,
    payload: dict[str, int],
    session: SessionDep,
) -> dict[str, object]:
    """Add experience to one vision."""
    try:
        vision = await vision_services.add_experience_to_vision(
            session,
            vision_id=vision_id,
            experience_points=payload.get("experience_points", 0),
        )
    except (LookupError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _vision_payload(vision)


@router.post("/{vision_id}/harvest")
async def harvest_vision(vision_id: UUID, session: SessionDep) -> dict[str, object]:
    """Harvest one vision when the underlying LifeOS rules allow it."""
    try:
        vision = await vision_services.harvest_vision(session, vision_id=vision_id)
    except (LookupError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _vision_payload(vision)
